sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs

```diff
@@ -2487,21 +2487,29 @@ internal virtual void SetNameFieldsIfNull()
 
                         do
                         {
-                            jsonResponse =
-                                await DataLakeRestClient.Path.SetAccessControlRecursiveAsync(
-                                    clientDiagnostics: ClientDiagnostics,
-                                    pipeline: Pipeline,
-                                    resourceUri: DfsUri,
-                                    mode: mode,
-                                    maxRecords: options.BatchSize,
-                                    version: Version.ToVersionString(),
-                                    acl: accessControlList,
-                                    async: async,
-                                    continuation: continuationToken,
-                                    forceFlag: options.ContinueOnFailure ? true : (bool?)default,
-                                    cancellationToken: cancellationToken)
-                                .ConfigureAwait(false);
-
+                            try
+                            {
+                                jsonResponse =
+                                    await DataLakeRestClient.Path.SetAccessControlRecursiveAsync(
+                                        clientDiagnostics: ClientDiagnostics,
+                                        pipeline: Pipeline,
+                                        resourceUri: DfsUri,
+                                        mode: mode,
+                                        maxRecords: options.BatchSize,
+                                        version: Version.ToVersionString(),
+                                        acl: accessControlList,
+                                        async: async,
+                                        continuation: continuationToken,
+                                        forceFlag: options.ContinueOnFailure ? true : (bool?)default,
+                                        cancellationToken: cancellationToken)
+                                    .ConfigureAwait(false);
+                            }
+                            catch (RequestFailedException storageRequestFailedException)
+                            when (!string.IsNullOrEmpty(continuationToken) &&
+                                storageRequestFailedException.Status >= Constants.HttpStatusCode.ServerError)
```

Comment:
What about 400s?  @kasobol-msft

Reply:
400s are expected at first call, so no continuation there.

Reply:
Btw. I'd also cap the status from the other side i.e. 500<=status<600.

One more thing to check is timeouts/cancellations

Reply:
Unless the customer passed a ContinuationToken?  And we should just catch/wrap all exceptions.  There are a lot of other things that can get thrown like TaskCanceledException.

Reply:
I think it's good idea. However, I'd consider not wrapping 4xx - that's clear signal that first api call was wrong - if we wrap it then one would need to inspect inner exception to which might be seen as overhead (not very harmful though).

Reply:
What gets returned when you don't have permissions to set an ACL?  I assume a 4XX.  I think an allow-list of 4XXs (or an allow-list of error codes) like 404 would be fine to not wrap, but I wouldn't let all 4XXs through on the first request.

Reply:
401 as far as I recall.

Reply:
@tg-msft Should we just wrap all the exceptions that come through? That way we don't have a complicated error handling logic and just document that they should check the innerException to verify whether or not they had a retriable error not. Also so that the customer isn't trying to catch `DataLakeAclChangeFailedException` and `RequestFailedException`. `DataLakeAclChangeFailedException` is a pretty generic name to cover this.

Reply:
With catching every type of exception it would be required that the custom Exception wraps around each possible exception type.
```
public class DataLakeAclChangeFailedException : Exception, ISerializable
    {
        public RequestFailedException InnerRequestFailedException { get; }

        public TaskCanceledException InnerTaskCanceledException { get; }

        public string ContinuationToken { get; }
 ...
```
Is there a better way of doing this, or should I go down with doing this? We could catch a regular `Exception` but then we would lose exception info like `RequestFailedException.Status` and `ErrorCode`, as well as the `TaskCanceledException.Task` and `CancellationToken` (right?)

Reply:
We should just use https://docs.microsoft.com/en-us/dotnet/api/system.exception.innerexception?view=netcore-3.1 instead of adding multiple exception properties.  You can set that in the .ctor when you create `DataLakeAclChangeFailedException`.

We could maybe write something like this to be extra helpful when people are trying to debug:

```C#
try
{
    // do work here ...
}
catch (RequestFailedException ex)
{
    throw new DataLakeAclChangeFailedException(
        $"An error occurred while recursively changing the access control list.  " +
        $"See the {nameof(ex.InnerException)} of type {ex.GetType().FullName} with " +
        $"{nameof(ex.Status)}={ex.Status} and {nameof(ex.ErrorCode)}={ex.ErrorCode} for more information.",
        ex);
}
catch (Exception ex)
{
    throw new DataLakeAclChangeFailedException(
        $"An error occurred while recursively changing the access control list.  " +
        $"See the {nameof(ex.InnerException)} of type {ex.GetType().FullName} for more information.",
        ex);
}
```

I'm 90% sure doing a `ToString` on a `DataLakeAclChangeFailedException` would also include the `InnerException`'s `ToString` as part of that, but we should double check to be sure.  If it doesn't, maybe we also hoist `InnerException.Message` up to `DataLakeAclChangeFailedException`'s `Message`.

I could go either way on catching all `RequestFailedException`s.  On the one hand, you haven't really started changing anything yet when the failure occurs  But that's not a distinction customers would easily grasp.  It might be easier to have just one exception they need to catch here.  I think always wrapping in `DataLakeAclChangeFailedException` might be a little better, but I'm fine with whatever you all decide is the best way to represent the service.

Reply:
let's wrap'em all.